datalad_next/runners/run.py

```diff
@@ -0,0 +1,166 @@
+"""
+This module provides a context-manager that executes a subprocess.
+
+The context manager supports the termination and/or killing of the subprocess,
+if it is executed with a not-None `terminate_time` and/or `kill_time`.
+
+The context manager supports generator- and non-generator-protocols.
+
+"""
+from __future__ import annotations
+
+import subprocess
+from collections.abc import Generator
+from contextlib import contextmanager
+from pathlib import Path
+from queue import Queue
+from subprocess import DEVNULL
+from typing import (
+    Any,
+    IO,
+)
+
+from datalad_next.runners import (
+    GeneratorMixIn,
+    Protocol,
+    ThreadedRunner,
+)
+
+
+def create_kill_wrapper(cls: type[Protocol]) -> type[Protocol]:
+    """ Extend `cls` to supports the "kill-interface"
+
+    This function creates a subclass of `cls` that contains the components
+    of the "kill-interface".  The two major components are amethod called `arm`,
+    and logic inside the timeout handler that can trigger a termination or
+    kill signal to the subprocess, if the termination time or kill time has
+    come.
+
```

Comment:
It is only used internally and a mixture between a real interface, i.e. it contains the method `arm`, and a behavior overloading, i.e. wrap `timeout` of the base-class to implement the kill-logic.

Since it is not a simple interface I put it in quotes.

I left it as it is for now. I think I might rather replace "kill-interface" with "termination and kill mechanisms" or simillar